certora/specs/ExternalCalls.spec

```
@@ -0,0 +1,30 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+// True when a CALL has been placed.
+persistent ghost bool hasExternalCall  {
+    init_state axiom hasExternalCall == false;
+}
+
+// True when a DELEGATECALL has been placed.
+persistent ghost bool hasDelegateCall {
+    init_state axiom hasDelegateCall == false;
+}
+
+hook CALL(uint g, address addr, uint value, uint argsOffset, uint argsLength, uint retOffset, uint retLength) uint rc {
+        hasExternalCall = true;
+}
+
+hook DELEGATECALL(uint g, address addr, uint argsOffset, uint argsLength, uint retOffset, uint retLength) uint rc {
+    hasDelegateCall = true;
+}
+
+// Check that the contract is reentrant safe as it makes no external call.
+invariant reentrancySafe()
+  !hasExternalCall;
+
+// Check that the contract makes no delegate call.
```

So actually I just checked and it seems like the implementation contract can make a delegate call when upgrading. This is not what I thought at first, and it goes as follows:
- proxy delegate calls implem with `upgradeToAndCall`
  - implem sets the new implementation
  - implem also delegate calls the new implmentation. This is what we use in deployment now, to make contract creation and initialization. See [here](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/69c8def5f222ff96f2b5beff05dfba996368aa79/contracts/proxy/ERC1967/ERC1967Utils.sol#L60-L76), it seems to be named "setup call"

And this is why you need to filter out `upgradeToAndCall` here. I think it makes sense to mention it:
```suggestion
// Check that the contract makes no delegate call, except during an upgrade when doing a setup call to the new implementation.
```

---

Implemented in https://github.com/morpho-org/morpho-token-upgradeable/pull/90/commits/b4f049bf904daddac992bf417f403d44039ea170.